enarx-srv-traits/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+//! The traits contained in this crate constitute a big state machine for
+//! lauching and managing a keep across multiple technologies. Launching
+//! a keep goes through multiple states, in this order:
+//!
+//!   1. Preparation
+//!   2. Building
+//!   3. Attesting
+//!   4. Launching
+//!   5. Running
+//!
+//! ## Preparation
+//!
+//! The purpose of the preparation state is to give the technology a view
+//! of all static binary pages and entry point before creating an actual
+//! keep. This is useful for technologies like SGX where we need to hash
+//! and sign the binary before creating the keep.
+//!
+//! A keep loader will start by creating a `Preparer` object by calling
+//! `Preparer::new()`. If the technology represented by this type is not
+//! supported by the system, it will return `None`.
+//!
+//! Once a `Preparer` is obtained, the keep manager will call `Loader::load()`
+//! to load all the pages of the static binary. Then finally, the manager
+//! will call `Preparer::next()` to specify the entry point and conclude the
+//! preparation state.
+//!
+//! ## Building
+//!
+//! The purpose of the building state is to actually create a full keep
+//! with all the pages actually loaded. The workflow is similar to the
+//! preparation state.
+//!
+//! ## Attestation
+//!
+//! The purpose of the attestation state is attest the keep to a remote
+//! entity (usually the tenant). This state proxies packets back and forth
+//! between the keep and the tenant. Note that the contents of attestation
+//! packets is defined by the specific technology in question.
+//!
+//! ## Launching
+//!
+//! The purpose of the launching state is to proxy communication between
+//! the tenant and the keep after attestation. Note that the intent of this
+//! state is to proxy integrity protected ciphertext. The manager should
+//! have no visibility into its contents.
+//!
+//! ## Running
+//!
+//! The purpose of the running state is to handle syscalls from the keep.
+
+use std::io::Result;
+
+use bitflags::bitflags;
+
+bitflags! {
+    /// Memory access permissions
+    pub struct Access: u8 {
+        /// Read access
+        const R = 1 << 0;
+
+        /// Write access
+        const W = 1 << 1;
+
+        /// Execute access
+        const X = 1 << 2;
+    }
+}
+
+/// A process which can either continue or terminate
+pub enum Continuation<N, D> {
+    /// The process is not complete
+    Next(N),
+
+    /// The process is complete
+    Done(D),
+}
+
+/// A syscall request
+pub enum SysCallReq {
+    None,
+}
+
+/// A syscall reply
+pub enum SysCallRep {
+    None,
+}
+
+/// A trait for supplying statically linked pages into the keep
+pub trait Loader {
+    fn load(&mut self, pages: &[u8], offset: u64, access: Access) -> Result<()>;
+}
+
+/// Prepares to build the keep
+pub trait Preparer: Sized + Loader {
```

Comment:
IIUC, this means that the type must implement both Sized and Loader. I guess I kind of understand why Loader is required but why is Sized required ? What happens if we don't specify Sized here ?

Reply:
@whitebrandy @lkatalin See the answer below.